Create a "TS basics" tut (#80)
tskit-dev/tutorials: basics.md

sample nodes will be on the tips of a tree: they represent genomes with a fully known
genetic sequence. Internal nodes (circles) represent ancestors in which a single DNA
sequence was duplicated (in forwards-time terminology) or in which multiple sequences
coalesced (in backwards-time terminology). There are 6 sample nodes, labelled $0..5$,
and also 6 non-sample nodes, labelled $6..11$, in the tree sequence above:

```{code-cell} ipython3
print("There are", ts.num_nodes, "nodes, of which", ts.num_samples, "are sample nodes")
```

### Edges

In an evolutionary tree, the connections between nodes are conventionally referred to as
"branches". In a tree sequence, we instead refer to directional connections between nodes
as _edges_, emphasizing that they usually persist across multiple trees. Each edge is
associated with a parent node ID and a child node ID. The time of the parent node must be
strictly greater than the time of the child node, and the difference in these times is
sometimes referred to as the "length" of the edge. Since trees in a tree sequence are
usually taken to represent marginal trees along a genome, as well as the time dimension

> heach edge also has a genomic _span_, defined by the edge's *left* and a *right* position

Reviewer: ```suggestion
each edge also has a genomic _span_, defined by the edge's *left* and a *right* position
```

sequence is defined using edges like $(0\rightarrow7)$ which spans the entire genome,
edges like $(10\rightarrow3)$ which only span the leftmost section of the genome, and
edges like $(11\rightarrow9)$ which span the rightmost part of the genome. Every time an
edge changes it creates a new tree; the location of the change is known as a _breakpoint_:

```{code-cell} ipython3
print("There are", ts.num_trees, "trees, with breakpoints", ts.breakpoints(as_array=True))
```
:::{note}
The tree sequence we have used here involves a tree with strictly bifurcating splits, but
this does not need to be the case for tree sequences in general. In particular, where
tree sequences have been created by the multiple-merger coalecent process, or in tree
sequences that have been inferred from real data, it is possible to have a parent node
with 3 or more children in a particular tree. Such multifucations are known as
*polytomies*.
:::

### Individuals and populations

Sometimes we know that one or more nodes belong to a particular _individual_. For example,
most organisms are diploid, so if we sequence the DNA from an individual we are likely to

> obtain two copies of each autosomal chromosome. The tree above could therefore represent

Reviewer: ```suggestion
obtain two copies of each autosomal chromosome. The tree with six sample nodes above could therefore represent
```

    ".node > .lab {font-size: 70%}"
    ".leaf > .lab {text-anchor: start; transform: rotate(90deg) translate(6px)}"
)
SVG(ts.draw_svg(
    node_labels=node_labels, style=styles, y_axis=True, y_gridlines=True, time_scale="rank"))
```

In the same way that nodes can be associated with a specific individual, nodes can also
be associated with a larger _population_. In this example, there is only one population
(to which all the nodes belong):

```{code-cell} ipython3
print(f"{ts.num_populations} population(s) defined:")
for population in ts.populations():
    print("*", population)
```

### Mutations and sites

In {program}`tskit`, _mutations_ occur at specific, defined _sites_. A mutation is
positioned above a particular node (specified by the node ID), occurs at a particular
site (specified by a site ID), and involves a change of allelic state to a defined
{attr}`~Mutation.derived_state`. A mutation therefore can be specified something like
this:

```{code-cell} ipython3
print(f"{ts.num_mutations} mutation(s) in the tree sequence:")
for mutation in ts.mutations():
    print("*", mutation)
```

The mutation can have a {attr}`~Mutation.time` or if, as in this case, the times of

> mutations in the tree sequence are unknown, all mutations can have the special value

Reviewer: ```suggestion
mutations in the tree sequence are unknown, all mutations can have the special NaN value
```

Instead, that is a property of the _site_ to which the mutation refers, in this case,
site ID 0 (which happens to be at position 257):

```{code-cell} ipython3
print(ts.site(0))  # For convenience, the Python API returns the mutations at this site too
```

In the plot above, since the the only mutation is above node 8 and has a
{attr}`~Mutation.derived_state` of "G", we know that the samples descending from node
8 in the first tree (sample genomes 2, 4, and 5) have a "G" at {attr}`~Site.position` 257,
while the others have the {attr}`~Site.ancestral_state` of "T". This means that Ada is
homozygous for "T", Bob is heterozygous "T|G", and Cat is homozygous for "G".
In other words the ancestral state and the details of any mutations at that site,
when coupled with the tree topology at the site {attr}`~Site.position` is sufficient to
define the allelic state possessed by each sample.

Note that even though the genome is 1000 base pairs long, the tree sequence only contains
a single site, because we usually only bother defining *variable* sites in a tree
sequence (e.g. positions seen in studies to have samples possessing different alleles at

> that genomic location). It is perfectly possible to have a site with no mutations

Reviewer: ```suggestion
that genomic location). It is perfectly possible to have a site with no mutations or silent mutations
```

SVG(ts.draw_svg(y_axis=True, y_gridlines=True, time_scale="rank", style=".mut {display: none}"))
```

### Tree changes, ancestral recombinations, and SPRs

Note that adjacent trees in this tree sequence differ by a single "tree edit" or SPR
(subtree-prune-and-regraft) operation: the subtree defined by node 8 in the first
tree has been pruned and regrafted onto the branch from 10 to 3, to create the
second tree.

The second and third trees have the same topology, but differ because their
ultimate coalesence happened in a different ancestor (this is easy to see in a simulation,
but hard to detect in real data). The SPR that created the third tree must involve pruning
either the subtree below node 7 or the subtree below node 9 and grafted it higher up on
the same lineage. In this case it turns out to be impossible to tell just from the tree
sequence which of those two SPRs actually happened in the simulation.

In general, each detectable recombination occurring in ancestral history results in a
single SPR in the tree sequence. If recombination breakpoints occurs at unique
positions (an "infinite sites" model of breakpoints), then the number of trees in a tree

> sequence equals the number of ancestral recombination events minus one. If recombinations

Reviewer: I'm either missing something here or this should be that the number of trees is equal to the number of recombinations plus one? :confused:

Author: Well spotted. Copied from the old workshop material, where it's presumably wrong too!

two reasons:

1. Many recombination events are undetectable from sequence data, and even if they are
   detectable, they can be logically impossible to place in the genealogy (as in the
   second SPR example above).
2. The number of recombination events in the genealogy can grow to dominate the total
   number of nodes in the total tree sequence, without actually contributing to the
   realised sequences in the samples. In other words, recombination nodes are redundant
   to the storing of genome data.

In other words, compared to an ARG, a standard tree sequence simply stores the trees
created by recombination events, rather than attempting to record the recombination
events themselves. Many (but as we have seen, not all) details of the actual
recombination events can be inferred from the changes in edges along a tree sequence.
In other words

> "an ARG encodes the events that occurred in the history of a sample,
> whereas a tree sequence encodes the outcome of those events"
> ([Kelleher _et al._, 2019](https://doi.org/10.1534/genetics.120.303253))

:::{todo}
Link to the ARG tutorial: https://github.com/tskit-dev/tutorials/issues/43
:::

### Tables

:::{todo}

> The underlying data in a tree sequence is best thought of as a set of tables. The API

Reviewer: ```suggestion
The underlying data in a tree sequence is stored as a set of tables. The API
```
